test/asynchronous/test_database.py

```diff
@@ -0,0 +1,778 @@
+        )
+        try:
+            db = self.client.pymongo_test
+            await db.command("count", "test")
+            with self.assertRaises(ExecutionTimeout):
+                await db.command("count", "test", maxTimeMS=1)
+            pipeline = [{"$project": {"name": 1, "count": 1}}]
+            # Database command helper.
+            await db.command("aggregate", "test", pipeline=pipeline, cursor={})
+            with self.assertRaises(ExecutionTimeout):
+                await db.command(
+                    "aggregate",
+                    "test",
+                    pipeline=pipeline,
+                    cursor={},
+                    maxTimeMS=1,
+                )
+            # Collection helper.
+            await db.test.aggregate(pipeline=pipeline)
+            with self.assertRaises(ExecutionTimeout):
+                await db.test.aggregate(pipeline, maxTimeMS=1)
+        finally:
+            await self.client.admin.command(
+                "configureFailPoint", "maxTimeAlwaysTimeOut", mode="off"
+            )
+
+    def test_with_options(self):
+        codec_options = DECIMAL_CODECOPTS
+        read_preference = ReadPreference.SECONDARY_PREFERRED
+        write_concern = WriteConcern(j=True)
+        read_concern = ReadConcern(level="majority")
+
+        # List of all options to compare.
+        allopts = [
+            "name",
+            "client",
+            "codec_options",
+            "read_preference",
+            "write_concern",
+            "read_concern",
+        ]
+
+        db1 = self.client.get_database(
+            "with_options_test",
+            codec_options=codec_options,
+            read_preference=read_preference,
+            write_concern=write_concern,
+            read_concern=read_concern,
+        )
+
+        # Case 1: swap no options
+        db2 = db1.with_options()
+        for opt in allopts:
+            self.assertEqual(getattr(db1, opt), getattr(db2, opt))
+
+        # Case 2: swap all options
+        newopts = {
+            "codec_options": CodecOptions(),
+            "read_preference": ReadPreference.PRIMARY,
+            "write_concern": WriteConcern(w=1),
+            "read_concern": ReadConcern(level="local"),
+        }
+        db2 = db1.with_options(**newopts)  # type: ignore[arg-type]
+        for opt in newopts:
+            self.assertEqual(getattr(db2, opt), newopts.get(opt, getattr(db1, opt)))
+
+
+class TestDatabaseAggregation(AsyncIntegrationTest):
+    def setUp(self):
+        self.pipeline: List[Mapping[str, Any]] = [
+            {"$listLocalSessions": {}},
+            {"$limit": 1},
+            {"$addFields": {"dummy": "dummy field"}},
+            {"$project": {"_id": 0, "dummy": 1}},
+        ]
+        self.result = {"dummy": "dummy field"}
+        self.admin = self.client.admin
+
+    async def test_database_aggregation(self):
+        cursor = await self.admin.aggregate(self.pipeline)
```

Comment:
You can use `async with await self.admin.aggregate(self.pipeline) as cursor:` and keep it similar to how the synchronous version was before.